tensorflow/contrib/layers/ln_kernels/layer_norm_fused_op.cc

```diff
@@ -0,0 +1,344 @@
+
+ public:
+
+  explicit LayerNormFusedOp(OpKernelConstruction* context)
+      : OpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("epsilon", &epsilon_));
+
+  }
+
+  void Compute(OpKernelContext* context) override {
+
+    const Tensor& input = context->input(0);
+    const Tensor& gamma = context->input(1);
+    const Tensor& beta = context->input(2);
+
+    OP_REQUIRES(context, input.dims() >= 2,
+                errors::InvalidArgument("input dimensions must be larger than 2D",
+                                        input.shape().DebugString()));
+    OP_REQUIRES(context, gamma.dims() == 1,
+                errors::InvalidArgument("gamma dimension must be 1D",
+                                        gamma.shape().DebugString()));
+    OP_REQUIRES(context, beta.dims() == 1,
+                errors::InvalidArgument("beta dimension must be 1D",
+                                        beta.shape().DebugString()));
+
+    const int32 last_dim = input.dims()-1;
+    const int32 depth = input.dim_size(last_dim);
+
+    OP_REQUIRES(
+        context, depth == gamma.dim_size(0),
+        errors::InvalidArgument("input depth and gamma must have the same size: ",
+                                depth, " vs ", gamma.dim_size(0)));
+    OP_REQUIRES(
+        context, depth == beta.dim_size(0),
+        errors::InvalidArgument("input depth and beta must have the same size: ",
+                                depth, " vs ", beta.dim_size(0)));
+
+    int32 n_slices = 1;
+    for (int i = 0; i < last_dim; ++i)
+    {
+        n_slices *= input.dim_size(i);
+    }
+
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context, context->allocate_output(0, input.shape(), &output));
+
+    VLOG(2) << "LayerNormFusedCustom: "
+            << "depth:"<<depth << ", " <<"n_slices:"<<n_slices;
+
+    //temporarily hard-coding warp_size for CUDA kernel.
+    const int warp_size = 32;
+    LayerNormFusedArgs args;
+    args.depth = depth;
+    args.n_slices = n_slices;
+    args.n_inputs = n_slices*depth;
+    args.epsilon = epsilon_;
+
+    if (depth<=warp_size)
+    {
+        int tmp_depth = depth;
+        int slice_size = 1;
+        while(tmp_depth >>=1)slice_size*=2;
+        args.slice_size = slice_size>=depth?slice_size:slice_size*2;
+    }else{
+        int slice_size = (depth/warp_size)*warp_size;
+        args.slice_size = slice_size>=depth?slice_size:slice_size+warp_size;
+    }
+    auto input_ptr = input.template flat<T>().data();
+    auto gamma_ptr = gamma.template flat<T>().data();
+    auto beta_ptr = beta.template flat<T>().data();
+    auto output_ptr = output->template flat<T>().data();
+
+    LaunchLayerNormFusedOp<Device, T>::launch(context, args, input_ptr,gamma_ptr,beta_ptr,
+                                             output_ptr
+                                             );
+  }
+
+ private:
+  float epsilon_;
+
+
+  TF_DISALLOW_COPY_AND_ASSIGN(LayerNormFusedOp);
+};
+
+
+
+
+
+REGISTER_KERNEL_BUILDER(
```

Comment:
Yes! I removed them so I could compile the kernel without using bazel and forgot to add them back before I send the pull request.